cdmtaskservice/input_file_locations.py

```diff
@@ -0,0 +1,93 @@
+'''
+Module that determines a file's expected path given a job input.
+'''
+
+from bidict import bidict
+import marisa_trie
+import math
+import os
+
+from cdmtaskservice.models import JobInput, S3File
+
+
+def determine_file_locations(job_input: JobInput) -> dict[S3File | str, str]:
+    """
+    Given job input, determine where files should be placed given a set of input roots.
+    
+    job_input - the job input.
+    
+    Returns a mapping of each file in the job input to the file path with any relevant input root
+    removed, or just the filename if no input roots match.
+    """
+    trie, pathseg_to_id = _setup_trie(job_input)
+    ret = {}
+    seen = {}
+    for i, f in enumerate(job_input.input_files):
+        fstr = f.file if isinstance(f, S3File) else f
+        # if the trie is empty or only contains "" could speed things up by not translating the
+        # path
+        idpath = _path_to_id_string(fstr, pathseg_to_id)
+        pref = [p for p in trie.prefixes(idpath) if p != idpath]
+        if pref:
+            pref = sorted(pref, key=lambda x: len(x))[-1]
+            loc = _id_string_to_path(idpath.removeprefix(pref + "/"), pathseg_to_id)
+        else:
+            loc = os.path.basename(fstr)
+        if loc in seen:
+            raise InputFileCollisionError(
+                f"Input files '{fstr}' at index {i} and '{seen[loc][0]}' "
+                + f"at index {seen[loc][1]} collide at path "
+                + f"'{job_input.params.input_mount_point}/{loc}'")
+        ret[f] = loc
+        seen[loc] = (fstr, i)
+    return ret
+
+
+def _setup_trie(job_input: JobInput) -> (marisa_trie.Trie, bidict[str, str]):
+    # Make a poor man's path trie using IDs to substitute for path segments.
+    # This seems dumb, there must be a better way
+    # other than creating a trie specifically for paths rather than strings, anyway
+    pathseg_to_id = {}
```

Comment:
```suggestion
    pathseg_to_id = bidict()
```

Reply:
It doesn't need to be a bidict at this point